### PR TITLE
add word2vec test for the new API

### PR DESCRIPTION
**Pull request by helinwang**

### CODE REVIEW

File: python/paddle/fluid/tests/book/no_test_word2vec_new_api.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    trainer = fluid.Trainer(
+        partial(inference_network, is_sparse),
+        optimizer=fluid.optimizer.SGD(learning_rate=0.001),
+        place=place,
+        event_handler=event_handler)
+    trainer.train(train_reader, 100)
+
+
+def infer(use_cuda, save_path):
+    params = fluid.Params(save_path)
```
Comment (sidgoyal78) [inline python/paddle/fluid/tests/book/no_test_word2vec_new_api.py]: maybe something like `fluid.params.read()` (since we have `trainer.params.save()` above?)

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        partial(inference_network, is_sparse),
+        optimizer=fluid.optimizer.SGD(learning_rate=0.001),
+        place=place,
+        event_handler=event_handler)
+    trainer.train(train_reader, 100)
+
+
+def infer(use_cuda, save_path):
+    params = fluid.Params(save_path)
+    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
```
Comment (cs2be) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Sorry didn't catch this before, but place is not used, it should be passed into fluid.Inferencer

File: python/paddle/fluid/tests/book/no_test_word2vec_new_api.py (hunk 1 above)
Comment (helinwang) [reply python/paddle/fluid/tests/book/no_test_word2vec_new_api.py]: Hmm, if we aim for symmetry, `params.save` is a method of a `params` instance, in the same time `fluid.params.load` is a static method, which is not very symmetric?

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py (hunk 2 above)
Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Thanks, done.

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                              act='sigmoid')
+    predict_word = fluid.layers.fc(input=hidden1, size=dict_size, act='softmax')
+    return predict_word
+
+
+def train_network():
+    next_word = fluid.layers.data(name='nextw', shape=[1], dtype='int64')
+    predict_word = inference_network()
+    cost = fluid.layers.cross_entropy(input=predict_word, label=next_word)
+    avg_cost = fluid.layers.mean(cost)
```
Comment (reyoung) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: The metrics, like error rate, auc, are not defined in this demo. How do we define a error rate in our new designed API?

By just adding fluid.layers.accury(input=predict_word, label=label) ? Should we fetch metrics by default?

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+            if avg_cost < 5.0:
+                trainer.params.save(save_path)
+                return
+            if math.isnan(avg_cost):
+                sys.exit("got NaN loss, training failed.")
+
+    trainer = fluid.Trainer(
+        partial(inference_network, is_sparse),
+        fluid.optimizer.SGD(learning_rate=0.001),
```
Comment (reyoung) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: If we passing the optimizer instance to a trainer, we cannot use learning rate schedule since the learning_rate is a variable of learning rate schedule method, which is designed by @jacquesqiao 

Perhaps, we can pass a lambda here.

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                word_dict, N))
+
+            if avg_cost < 5.0:
+                trainer.params.save(save_path)
+                return
+            if math.isnan(avg_cost):
+                sys.exit("got NaN loss, training failed.")
+
+    trainer = fluid.Trainer(
+        partial(inference_network, is_sparse),
```
Comment (reyoung) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: `partial(train_network, is_sparse)` ?

Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Yes, `fluid.Trainer`'s first argument should be a function without any arguments. `partial(train_network, is_sparse)` binds `is_sparse` to `train_network`, creating a function without any argument.

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py (hunk 4 above)
Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Cool! I realized that we may need to freeze parameter when reading your GAN example
Another possibility is we pass in a optimize function?

```Python
def optimize(loss):
    opt = fluid.optimizer. SGD(learning_rate=fluid.layers.fill_constant(1e-3)))
    opt.minimize(loss=loss)
    return opt.minimize(
        loss=dg_loss,
        startup_program=startup_program,
        parameter_list=[
            p.name for p in g_program.global_block().all_parameters()
        ])

trainer = fluid.Trainer(..., optimize=optimize)
```
